packages/flutter_test/lib/src/widget_tester.dart

```diff
@@ -158,6 +158,19 @@ void expectSync(dynamic actual, dynamic matcher, {
   test_package.expect(actual, matcher, reason: reason);
 }
 
+/// Just like [expect], but returns a [Future] that completes when the matcher
+/// has finished matching.
+///
+/// If the matcher fails asynchronously, that failure is piped to the returned
+/// future where it can be handled by user code.
+Future<void> expectLater(dynamic actual, dynamic matcher, {
+  String reason,
+  dynamic skip, // true or a String
+}) {
+  TestAsyncUtils.guardSync();
```

Comment:
Should this be wrapped in a guard, rather than just being a synchronous guard? Or is it ok if `expectLater` isn't awaited?

Reply:
Turns out this does need to just be `guardSync()`.  Otherwise, we run into other async barriers while we're waiting for the matcher to complete.  e.g.:

https://github.com/flutter/flutter/blob/8860627b635c6f54a359e953f14258e7accc897b/packages/flutter_test/lib/src/binding.dart#L485

Reply:
Sent https://github.com/flutter/flutter/pull/17085 to fix this.